drivers/sensor/nordic/npm2100_vbat/npm2100_vbat.c

```diff
@@ -0,0 +1,563 @@
+	/* Fetch previous DPS duration result before triggering new one.The time it takes to get
+	 * the DPS duration depends on many factors and cannot be predicted here.
+	 */
+	ret = i2c_reg_read_byte_dt(&config->i2c, BOOST_DPSDUR, &data->dpsdur);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return i2c_reg_write_byte_dt(&config->i2c, BOOST_START, 2U);
+}
+
+static int npm2100_vbat_attr_get(const struct device *dev, enum sensor_channel chan,
+				 enum sensor_attribute attr, struct sensor_value *val)
+{
+	const struct npm2100_vbat_config *const config = dev->config;
+
+	if (attr == SENSOR_ATTR_FEATURE_MASK) {
+		struct adc_config *adc_cfg = adc_cfg_get(dev, chan);
+
+		if (!adc_cfg) {
+			return -EINVAL;
+		}
+
+		*val = (struct sensor_value){.val1 = adc_cfg->enabled, .val2 = 0};
+
+		return 0;
+	}
+
+	if (attr == SENSOR_ATTR_OVERSAMPLING) {
+		struct adc_config *adc_cfg = adc_cfg_get(dev, chan);
+
+		if (!adc_cfg) {
+			return -EINVAL;
+		}
+
+		*val = (struct sensor_value){
+			.val1 = FIELD_GET(ADC_CONFIG_AVG_MASK, adc_cfg->config), .val2 = 0};
+
+		return 0;
+	}
+
+	/* Delay of vbat ADC measurement */
+	if ((chan == SENSOR_CHAN_GAUGE_VOLTAGE) &&
+	    (attr == (enum sensor_attribute)SENSOR_ATTR_NPM2100_ADC_DELAY)) {
+		struct npm2100_vbat_data *const data = dev->data;
+		struct adc_config *adc_cfg = adc_cfg_get(dev, chan);
+		int32_t val_usec;
+
+		if (!adc_cfg) {
+			return -ENOENT;
+		}
+
+		if (FIELD_GET(ADC_CONFIG_MODE_MASK, adc_cfg->config) == ADC_CONFIG_MODE_INS_VBAT) {
+			/* Instant measurement */
+			return sensor_value_from_micro(val, 0);
+		}
+
+		int ret = linear_range_get_value(&adcdelay_range, data->vbat_delay, &val_usec);
+
+		if (ret < 0) {
+			return ret;
+		}
+
+		return sensor_value_from_micro(val, val_usec);
+	}
+
+	for (int idx = 0; idx < ARRAY_SIZE(npm2100_attr); idx++) {
+		if ((npm2100_attr[idx].chan == chan) && (npm2100_attr[idx].attr == attr)) {
+
+			int32_t val_mv;
+			uint8_t reg_data;
+
+			int ret = i2c_reg_read_byte_dt(&config->i2c, npm2100_attr[idx].reg,
+						       &reg_data);
+
+			if (ret < 0) {
+				return ret;
+			}
+
+			reg_data = FIELD_GET(npm2100_attr[idx].reg_mask, reg_data);
+
+			ret = linear_range_get_value(npm2100_attr[idx].range, reg_data, &val_mv);
+			if (ret < 0) {
+				return ret;
+			}
+
+			val->val1 = val_mv / 1000000;
+			val->val2 = val_mv % 1000000;
+
+			return 0;
+		}
+	}
+
+	return -ENOTSUP;
+}
+
+static int npm2100_vbat_attr_set(const struct device *dev, enum sensor_channel chan,
+				 enum sensor_attribute attr, const struct sensor_value *val)
+{
+	const struct npm2100_vbat_config *const config = dev->config;
+	int ret;
+
+	/* ADC sampling feature masks to enable individual measurements */
+	if (attr == SENSOR_ATTR_FEATURE_MASK) {
+		struct adc_config *adc_cfg = adc_cfg_get(dev, chan);
+
+		if (!adc_cfg) {
+			return -EINVAL;
+		}
+
+		adc_cfg->enabled = val->val1 == 0 ? false : true;
+
+		return 0;
+	}
+
+	/* Delay of vbat ADC measurement */
+	if ((chan == SENSOR_CHAN_GAUGE_VOLTAGE) &&
```

Comment:
wondering if this should be a fuel gauge driver?

Reply:
The npm2100 hardware only measures voltages, which I think fits better as a sensor device driver. To provide more full fledged fuel gauge functionality, additional software is required. We do have this kind of software in NCS/sdk-nrfxlib, and we are considering supporting the zephyr fuel gauge API, but I think that is a larger task outside the scope of this driver